src/main/scala/sigmastate/eval/CostingRules.scala

```diff
@@ -197,47 +197,53 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
 
-    def constantSizeProperyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] =
-      withConstantSize(prop(obj.value), opCost(costOfArgs, selectFieldCost))
+    def constantSizePropertyAccess[R](prop: Rep[T] => Rep[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      withConstantSize(value, opCost(value, costOfArgs, selectFieldCost))
+    }
 
-    def knownSizeProperyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] =
-      RCCostedPrim(prop(obj.value), opCost(costOfArgs, selectFieldCost), size)
+    def knownSizePropertyAccess[R](prop: Rep[T] => Rep[R], size: RSize[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      RCCostedPrim(value, opCost(value, costOfArgs, selectFieldCost), size)
+    }
 
-    def knownLengthCollProperyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
-      mkCostedColl(prop(obj.value), len, opCost(costOfArgs, selectFieldCost))
+    def knownLengthCollPropertyAccess[R](prop: Rep[T] => Rep[Coll[R]], len: Rep[Int]): Rep[CostedColl[R]] =
+      mkCostedColl(prop(obj.value), len, opCost(obj.value, costOfArgs, selectFieldCost))
 
-    def digest32ProperyAccess(prop: Rep[T] => Rep[Coll[Byte]]): Rep[CostedColl[Byte]] =
-      knownLengthCollProperyAccess(prop, CryptoConstants.hashLength)
+    def digest32PropertyAccess(prop: Rep[T] => Rep[Coll[Byte]]): Rep[CostedColl[Byte]] =
+      knownLengthCollPropertyAccess(prop, CryptoConstants.hashLength)
 
-    def groupElementProperyAccess(prop: Rep[T] => Rep[GroupElement]): RCosted[GroupElement] =
-      knownSizeProperyAccess(prop, SizeGroupElement)
+    def groupElementPropertyAccess(prop: Rep[T] => Rep[GroupElement]): RCosted[GroupElement] =
+      knownSizePropertyAccess(prop, SizeGroupElement)
 
-    def bigIntProperyAccess(prop: Rep[T] => Rep[BigInt]): RCosted[BigInt] =
-      knownSizeProperyAccess(prop, SizeBigInt)
+    def bigIntPropertyAccess(prop: Rep[T] => Rep[BigInt]): RCosted[BigInt] =
+      knownSizePropertyAccess(prop, SizeBigInt)
 
-    def defaultProperyAccess[R](prop: Rep[T] => Rep[R], propSize: RSize[T] => RSize[R]): RCosted[R] =
-      RCCostedPrim(prop(obj.value), opCost(costOfArgs, selectFieldCost), propSize(obj.size))
+    def defaultPropertyAccess[R](prop: Rep[T] => Rep[R], propSize: RSize[T] => RSize[R]): RCosted[R] = {
+      val value = prop(obj.value)
+      RCCostedPrim(value, opCost(value, costOfArgs, selectFieldCost), propSize(obj.size))
+    }
 
-    def defaultOptionProperyAccess[R: Elem](prop: Rep[T] => ROption[R], propSize: RSize[T] => RSize[WOption[R]], itemCost: Rep[Int]): RCostedOption[R] = {
+    def defaultOptionPropertyAccess[R: Elem](prop: Rep[T] => ROption[R], propSize: RSize[T] => RSize[WOption[R]], itemCost: Rep[Int]): RCostedOption[R] = {
       val v = prop(obj.value)
       val s = propSize(obj.size)
-      RCCostedOption(v, SOME(itemCost), asSizeOption(s).sizeOpt, opCost(costOfArgs, selectFieldCost))
```

Comment:
@greenhat v should be used in opCost. Why it was not caught by asserts?

Reply:
Thanks for spotting this one! I failed at `assert` in `special`(checked the wrong cost).